### DISCUSSION TITLE
Functionality for splitting into components and listing component ingredients

### DISCUSSION THREAD

**Discussion by BlackLotus** [cooklang/spec#72] [Ideas]
So maybe relevant discussions are https://github.com/cooklang/spec/discussions/55, https://github.com/cooklang/spec/discussions/59 and #62
The main idea behind this is time management/planning and mise en place.
So when cooking you often have different component for instance you can have your sauce, your main dish and the side dishes. You could accomplish this through something like includes but when for instance baking this gets noticeably harder.

When baking you can have your sour dough, scald, soaker (not sure if it's really called that it would be Quellstück and Brühstück in german https://de.wikipedia.org/wiki/Quellst%C3%BCck), mother dough (Vorteig) and the main dough (Hauptteig)  and those are not really separable from the recipe (there are many MANY other "sub recipes" that would benefit sub recipes)
Ok this was really much and I was pretty unclear so here is what I propose.

**Splitting a recipe in sub recipes**
This could be done like described in #59 you could do `# Part 1` and `# Part 2` and seperate it this way
This could really play along well with #62

**Tracking preparation time**
So when cooking some things take longer or must be prepped in advance. Adding timestamps or the length of time it takes to prepare/must sit
`# Mother Dough $0$5m$12h$` for instance coud add a Sub-Component (mother dough) that takes place at the beginning of the process (0), takes 5min to prepare (5m) and has to rest for 12hours (12h). (the syntax for this is just a placeholder I don't have any good ideas yet)

**Tracking ingredients**
Like this you could create ingredient lists for the sub-products

---
To not muddle the water with features already discussed elsewhere I will try to reproduce one of my recipe entries for rye-wheat rolls (I feel dirty calling them that...)

**Comment by BlackLotus** [comment]
`NOTE: Baker percentage would be a nice feature as well. I was too lazy to actually calculate the % for this example`
`NOTE2: (the numbers behind the flour are a german thing)`
`NOTE3: The recipe is based on a recipe of Lutz Geißler https://www.ploetzblog.de/ I don't know which of his recipes it is thought`
`Roggenmischbrötchen`
# Planning
| Step | Day | Time |
| ------| ------| -----|
| Sour dough | Today | 21:00 |
| Main dough | Tomorrow | 11:00 |
| Form dough | Tomorrow | 12:00 |
| Preheat oven | Tomorrow | 13:15 |
| Bake | Tomorrow | 14:00 |

# Sour Dough
| Amount | Ingredient | Baker percentage|
|--|--|--|
| 150g | Rye flour 997 | % |
| 150g | Water | % |
| 5g | Salt | % |
| 30g | Starter | % |

Mix everything let rest for 14 hours.

# Main dough
| Amount | Ingredient | Baker percentage|
|--|--|--|
||Sour dough|
|140g|Rye Flour 997||
|130g| Wheat Flour 550||
|150g| Water (50C)||
|6g|Salt|
|9g| Butter (room temp)|
|15g|barley malt|

Mix for **5min** on lowest speed, mix **5min** on medium speed until you get an elastic dough.

Let dough rest for **1h**.

Form dough on floured surface.

Let rest for **3h** on floured cloth cover with cloth.

Bake at **230C** for **20min**.

**Reply by BlackLotus** [reply]
Note that this is probbly a bad copy of the original recipe. If you want to replicate this you should probably visit the ploetzblog and look for the recipe.
His website is really nice (and he is also very nice) https://www.ploetzblog.de/2022/02/26/ankunftsbrot-dinkel-roggen-brot/

I got many other ideas based on posting his recipe and I will start other discussions in hopes of improving the spec.

**Comment by dubadub** [comment]
@BlackLotus thanks for your idea. I'm checking that I understood everything correctly. 

You're proposing to introduce a composable recipe. The main consequence of this is that each sub-recipe is rendered like a separate recipe. Is that right? 

Also, you outlined the context very well and proposed a _solution_, but it's not really clear for me what it the _problem_ it solves. For example, problem can be that we want to cook parts of a recipe in parallel or something else. Maybe you can think what changes in the UI will it drive?

**Comment by wpearse** [comment]
Hello, I apologize if I am stepping on any toes, but many of our recipes are laid out in such a way that ingredients are listed in two distinct sections. I believe this is relevant to the discussion here.

For example, one of our recipes (on paper) reads:

```
Chicken Ingredients
- 1 tsp dried thyme
- 1 tsp smoked paprika
- 1 tsp onion powder
- 1 tsp garlic powder
- 2 tbsp butter, melted
- ½ tsp dried oregano
- 300 g chicken thighs, skinless, boneless

Rice Ingredients
- 1 brown onion, finely chopped
- 3 portobello mushrooms, finely sliced
- 4 cloves of garlic, minced
- 1 carrot, finely chopped
- 1½ C chicken stock, warmed
- 2 tbsp butter
- 1½ C long-grain white rice
```

We think that we can express many of our recipes in Cooklang as-is, but we haven't figured out how to deal with these ingredients (e.g. butter, in this recipe) that appear as two measures.

In print, right now, the recipe refers to the "first measure of butter" and the "second measure of butter" (this is the language "Hello Fresh" uses) ... and in some cases the method would be split into two (or more) sections (I have seen some discussion around composable/sub-recipes).

We were considering something like the following:

```
Preheat oven to bake 180°C. In a shallow bowl, combine ... @melted butter{0:2%tbsp}, salt and pepper. Add the @chicken{300%g} and coat each piece evenly with the seasoning.
...
Lightly oil a 25 x 35 cm dish. Add the @onion, ..., and @melted butter{1:2%tbsp} into the dish. Mix well to combine. Stir in the @rice.
```

Where the syntax might become `@ingredient{measure:quantity%unit}`. Of course, we would then need some way of indexing/naming the ingredient sections (measures 0 are "Chicken Ingredients", measures 1 are "Rice Ingredients").

In some way our "problems" might be solved by #59, too.

I hope this helps.

**Comment by dubadub** [comment]
Relevant proposal for spec change https://github.com/cooklang/spec/pull/104.

**Comment by pubmania** [comment]
@wpearse 
>we haven't figured out how to deal with these ingredients (e.g. butter, in this recipe) that appear as two measures.

The way I show ingredients with multiple measures is by displaying them under ingredients as separate entries and the step they appeared in so long as it was accompanied by a measure. I think that is an advantage because it allows me to keep two measures ready and not put the ingredient in incorrect order.

For example:

![image](https://github.com/user-attachments/assets/10824438-9f92-4ad4-8920-96339a3a77f5)

**Reply by dubadub** [reply]
Good idea. Depending on a context where we use that we can split ingredients (when cooking) or combine them (when shopping).
